### PR TITLE
feat(nuxt): add `dedupe` option for data fetching composables

### PR DESCRIPTION
**Pull request by theguriev**
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Added dedupe option to allow deduplicating requests to be used only once at the time. It is super useful in such cases:

```vue
<script setup lang="ts">
const { data, pending, error, refresh } = useAsyncData(
  "test",
  () =>
    new Promise((resolve) => {
      console.log("fetching data...");
      setTimeout(() => {
        resolve("Hello World");
      }, 1000);
    }), {
      dedupe: true
    }
);
useAsyncData(
  "test",
  () =>
    new Promise((resolve) => {
      console.log("fetching data...");
      setTimeout(() => {
        resolve("Hello World");
      }, 1000);
    }), {
      dedupe: true
    }
);
useAsyncData(
  "test",
  () =>
    new Promise((resolve) => {
      console.log("fetching data...");
      setTimeout(() => {
        resolve("Hello World");
      }, 1000);
    }), {
      dedupe: true
    }
);
useAsyncData(
  "test",
  () =>
    new Promise((resolve) => {
      console.log("fetching data...");
      setTimeout(() => {
        resolve("Hello World");
      }, 1000);
    }), {
      dedupe: true
    }
);
useAsyncData(
  "test",
  () =>
    new Promise((resolve) => {
      console.log("fetching data...");
      setTimeout(() => {
        resolve("Hello World");
      }, 1000);
    }), {
      dedupe: true
    }
);
useAsyncData(
  "test",
  () =>
    new Promise((resolve) => {
      console.log("fetching data...");
      setTimeout(() => {
        resolve("Hello World");
      }, 1000);
    }), {
      dedupe: true
    }
);
</script>

<template>
  <!-- Edit this file to play around with Nuxt but never commit changes! -->
  <div>Nuxt 3 Playground {{ data }}</div>
</template>

<style scoped></style>

``` 

I'm using something similar already in real prod project. But! I had to build another hook on top of that. It looks like this...

```vue
/* eslint-disable indent */

import { AsyncDataOptions } from "#app";

import { v1, ExtractFromAPI, PossibleMethods, Parameters as Params, replacePathParameters } from "@netgame/openapi";
import { NitroFetchOptions, NitroFetchRequest } from "nitropack";
import { hash } from "ohash";

import useAsync from "./useAsync";
import type { KeysOf } from "./useAsync";

export const dedupePromises = new Map();

const useAsyncFetch = <
	Path extends keyof v1.paths,
	Method extends keyof v1.paths[Path],
	Response extends ExtractFromAPI<v1.paths, Path, Method>,
	DataE = Error,
	DataT = Response,
	PickKeys extends KeysOf<DataT> = KeysOf<DataT>,
	DefaultT = null
>({
	path,
	method,
	options,
	fetchOptions
}: {
	path: Path;
	method: Method;
	options?: AsyncDataOptions<Response, DataT, PickKeys, DefaultT> & { cached?: true; key?: string };
	fetchOptions?:
		| (NitroFetchOptions<NitroFetchRequest> & Params<v1.paths[Path][Method]>)
		| (() => NitroFetchOptions<NitroFetchRequest> & Params<v1.paths[Path][Method]>);
}) => {
	const { path: parametersPaths } = (fetchOptions || {}) as unknown as { path: Record<string, string> };
	const headers = useRequestHeaders();
	const isQuasarClient = isClient && typeof window.__Q_META__ === "object";
	const baseURL = isQuasarClient ? undefined : "/api";
	const endpoint = replacePathParameters(path, parametersPaths);
	const asyncData = useAsync<Response, DataE, DataT, PickKeys, DefaultT>(
		options?.key || endpoint,
		() => {
			const requestParams = {
				baseURL,
				headers,
				...(typeof fetchOptions === "function" ? fetchOptions() : fetchOptions),
				method: method as PossibleMethods | undefined
			};

			const hashStr = hash({
				endpoint: options?.key || endpoint,
				requestParams
			});

			if (dedupePromises.has(hashStr)) {
				return dedupePromises.get(hashStr);
			}
			const promise = $fetch<Response>(endpoint, requestParams);

			dedupePromises.set(hashStr, promise);

			promise.finally(() => {
				dedupePromises.delete(hashStr);
			});

			return promise;
		},
		options
	);

	return asyncData;
};

export default useAsyncFetch;
``` 
So I think it would be cool to have it internally 😇 Let me know please if you see some issues. That's my first PR in Nuxt. 🥇 
### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Review by danielroe — COMMENTED** [review]
This is a nice idea.

Note that it introduces a use of `dedupe` that is the _opposite_ from the current usage (as an option to `refresh`). `refresh({ dedupe: false })` means 'do not _cancel_ existing requests in favour of this new one'.

But _this_ `dedupe: true` means `do not make any new requests if there is an existing pending request.

I am not in fact sure that my original choice of `dedupe` as the option for refresh was the most semantically correct but I think the usage should align between option to `useAsyncData` and the option to `refresh()`.

I would propose making `dedupe` accept these options:
- `'cancel'` (or `true`) - cancels existing requests when a new one is made <- default option
- `'defer'` (or `false`) - does not make new requests at all if there is a pending request

We would mark `true` or `false` types as deprecated and remove at a type level in a future minor.

Suggests welcome from @nuxt/core on naming choices here. 🙏

**Comment by theguriev** [comment]
> This is a nice idea.
> 
> Note that it introduces a use of `dedupe` that is the _opposite_ from the current usage (as an option to `refresh`). `refresh({ dedupe: false })` means 'do not _cancel_ existing requests in favour of this new one'.
> 
> But _this_ `dedupe: true` means `do not make any new requests if there is an existing pending request.
> 
> I am not in fact sure that my original choice of `dedupe` as the option for refresh was the most semantically correct but I think the usage should align between option to `useAsyncData` and the option to `refresh()`.
> 
> I would propose making `dedupe` accept these options:
> 
> * `'cancel'` (or `true`) - cancels existing requests when a new one is made <- default option
> * `'defer'` (or `false`) - does not make new requests at all if there is a pending request
> 
> We would mark `true` or `false` types as deprecated and remove at a type level in a future minor.
> 
> Suggests welcome from @nuxt/core on naming choices here. 🙏

Thank you! Yes it allows you to use the fetch anywhere without worrying about it being duplicated. Very useful in large companies when it's hard to keep track. 

I updated my PM but not sure I understood you correctly. Please take a look when you get a chance. Thank you so much! ❤️

File: packages/nuxt/src/app/composables/asyncData.ts [hunk 1]
```diff
@@ -172,7 +180,7 @@ export function useAsyncData<
 
   asyncData.refresh = asyncData.execute = (opts = {}) => {
     if (nuxt._asyncDataPromises[key]) {
-      if (opts.dedupe === false) {
+      if (isDefer(opts.dedupe) || isDefer(options.dedupe)) {
```
Comment (danielroe) [inline packages/nuxt/src/app/composables/asyncData.ts:183]: Let's allow overriding the global option.

```suggestion
      if (isDefer(opts.dedupe ?? options.dedupe)) {
```

**Review by danielroe — COMMENTED** [review]
Would you add a test in this section (that is, the 'describe' block) of the runtime composables test? 🙏 

https://github.com/nuxt/nuxt/blob/6e44b1b6b6611cc0f163a71308b74a0af7491d31/test/nuxt/composables.test.ts#L130-L147

**Comment by theguriev** [comment]
> Would you add a test in this section (that is, the 'describe' block) of the runtime composables test? 🙏
> 
> https://github.com/nuxt/nuxt/blob/6e44b1b6b6611cc0f163a71308b74a0af7491d31/test/nuxt/composables.test.ts#L130-L147

ofc. Here it is https://github.com/nuxt/nuxt/pull/24564/commits/8aa08d406066b9a747f0c501d13cf1cb3881e53f

Let me know please if I can improve it somehow. Thank you! :)

**Comment by Jamie4224** [comment]
Love this functionality! I implemented this by storing the useFetch in a pinia store, but for bigger projects with a lot of data sources, this is better!

**Review by danielroe — APPROVED** [review]

**Pull request merged**